### PR TITLE
[marshal] Add EH callback for native-to-managed wrapper

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by ludovic-henry** [comment]
cc @rolfbjarne

**Comment by ludovic-henry** [comment]
failed tests are `MonoTests.System.IO.FileInfoTest.Delete_Directory`

**Comment by ludovic-henry** [comment]
@kumpera could you please review?

**Comment by ludovic-henry** [comment]
build

File: mono/metadata/marshal.c [hunk 1]
```diff
@@ -8169,13 +8196,39 @@ mono_marshal_emit_managed_wrapper (MonoMethodBuilder *mb, MonoMethodSignature *i
...
+	/* throw e; */
+	mono_mb_emit_ldloc (mb, e_local);
+	mono_mb_emit_byte (mb, CEE_THROW);
+
+	/* } [endif] */
+	mono_mb_patch_branch (mb, rethrow_pos);
+
+	/* ex = mono_gc_handle_new (e, true); */
+	mono_mb_emit_ldloc (mb, e_local);
+	mono_mb_emit_icon (mb, 1);
```
Comment (kumpera) [inline mono/metadata/marshal.c]: No need to be pinned, the callback must switch to running to poke into managed memory anyways.

File: mono/metadata/marshal.c [hunk 2]
```diff
@@ -11817,3 +11880,22 @@ mono_marshal_free_dynamic_wrappers (MonoMethod *method)
...
+
+static void
+mono_marshal_ftnptr_eh_callback (guint32 gchandle)
+{
+	g_assert (ftnptr_eh_callback);
+	ftnptr_eh_callback (gchandle);
+}
+
+void
+mono_install_ftnptr_eh_callback (MonoFtnPtrEHCallback callback)
```
Comment (kumpera) [inline mono/metadata/marshal.c]: This must include documentation.

File: mono/metadata/marshal.h [hunk 3]
```diff
@@ -610,6 +610,11 @@ mono_mb_create_and_cache_full (GHashTable *cache, gpointer key,
 							   MonoMethodBuilder *mb, MonoMethodSignature *sig,
 							   int max_stack, WrapperInfo *info, gboolean *out_found);
 
+typedef void (*MonoFtnPtrEHCallback) (guint32 gchandle);
+
+void
```
Comment (kumpera) [inline mono/metadata/marshal.h]: This must be MONO_API.

**Comment by ludovic-henry** [comment]
@monojenkins merge

**Pull request merged**